agent.go

```diff
@@ -305,6 +309,58 @@ func (s *sandbox) teardownSharedPidNs() error {
 	return nil
 }
 
+func (s *sandbox) listenToUdevEvents() {
+	uEvHandler, err := uevent.NewHandler()
+	if err != nil {
+		return
+	}
+	defer uEvHandler.Close()
+
+	fieldLogger := agentLog.WithField("subsystem", "udevlistener")
+
+	for {
+		uEv, err := uEvHandler.Read()
+		if err != nil {
+			fieldLogger.Error(err)
+			continue
+		}
+
+		fieldLogger = fieldLogger.WithFields(logrus.Fields{
+			"uevent-action":    uEv.Action,
+			"uevent-devpath":   uEv.DevPath,
+			"uevent-subsystem": uEv.SubSystem,
+			"uevent-seqnum":    uEv.SeqNum,
+			"uevent-devname":   uEv.DevName,
+		})
+
+		// Check if device hotplug event results in a device node being created.
+		if uEv.DevName != "" && uEv.Action == "add" && strings.HasPrefix(uEv.DevPath, rootBusPath) {
+			s.Lock()
+
+			// Add the device node name to the pci device map.
+			s.pciDeviceMap[uEv.DevPath] = uEv.DevName
+
+			// Notify watchers that are interested in the udev event.
+			// Close the channel after watcher has been notified.
+			var matchedWatchers []string
+			for devPCIAddress, ch := range s.deviceWatchers {
+				if ch != nil && strings.HasPrefix(uEv.DevPath, filepath.Join(rootBusPath, devPCIAddress)) {
+					ch <- uEv.DevName
+					close(ch)
+					matchedWatchers = append(matchedWatchers, uEv.DevName)
```

Comment:
I think you can remove here to the watchers that have been notified, right? , hence you don't need `var matchedWatchers []string`

Reply:
I dont think it is safe to delete an entry in the map while looping through it.

Reply:
ohh, you're right  :smile:

Reply:
@devimc I just came across documentation that mentions that it is safe to delete from a a map while iterating through it. So now I am doing this within the loop itself.